pkg/conf/conf.go

```diff
@@ -170,33 +174,101 @@ func GetHarvestLogPath() string {
 	return logPath
 }
 
+func GetHarvestPidPath() string {
+	var pidPath string
+	if pidPath = os.Getenv("HARVEST_PIDS"); pidPath == "" {
+		pidPath = "/var/run/harvest/"
+	}
+	return pidPath
+}
+
+func main() {
+	configFp := "/home/rahulg2/code/github/harvest/harvest.yml"
+	if Config == (HarvestConfig{}) {
+		LoadHarvestConfig(configFp)
+	}
+
+	for k, _ := range *Config.Pollers {
+		GetPrometheusExporterPorts(k, configFp)
+	}
+}
+
 /*
 GetPrometheusExporterPorts returns port configured in prometheus exporter for given poller
 If there are more than 1 exporter configured for a poller then return string will have ports as comma seperated
 */
-func GetPrometheusExporterPorts(p *node.Node, configFp string) (string, error) {
-	var port string
-	exporters := p.GetChildS("exporters")
-	if exporters != nil {
-		exportChildren := exporters.GetAllChildContentS()
-		definedExporters, err := GetExporters(configFp)
-		if err != nil {
-			return "", err
-		}
-		for _, ec := range exportChildren {
-			promNode := definedExporters.GetChildS(ec)
-			if promNode == nil {
-				fmt.Printf("poller [%s] specified exporter [%s] that does not exist\n", p.GetNameS(), ec)
-				continue
+func GetPrometheusExporterPorts(pollerName string, configFp string) (int, error) {
+	var port int
+
+	if Config == (HarvestConfig{}) {
+		LoadHarvestConfig(configFp)
+	}
+	if len(promPortRangeMapping) == 0 {
+		LoadPrometheusExporterPortRangeMapping(configFp)
+	}
+	exporters := (*Config.Pollers)[pollerName].Exporters
+
+	if exporters != nil && len(*exporters) > 0 {
+		for _, e := range *exporters {
+			exporter := (*Config.Exporters)[e]
+			if *exporter.Type == "Prometheus" {
+				if exporter.PortRange != nil {
+					//fmt.Println(exporter.PortRange)
+					ports := promPortRangeMapping[e]
+					for p, _ := range ports {
+						if util.CheckPortAvailable(*exporter.Addr, strconv.Itoa(p)) {
+							port = p
+							break
+						}
+					}
+					for k, _ := range ports {
+						if k == port {
+							// remove ports which are already used
+							delete(ports, k)
+							break
+						}
+					}
+					fmt.Printf("chosen port %d \n", port)
+					return port, nil
+				} else if *exporter.Port != 0 {
+					fmt.Printf("port---- %d \n", *exporter.Port)
+					port = *exporter.Port
+					return port, nil
+				}
 			}
-			exporterType := promNode.GetChildContentS("exporter")
-			if exporterType == "Prometheus" {
-				currentPort := definedExporters.GetChildS(ec).GetChildContentS("port")
-				port = currentPort
+			continue
+		}
+
+	}
+	return port, errors.New(errors.ERR_CONFIG, "No free port found for poller "+pollerName)
+}
+
+var promPortRangeMapping = make(map[string]map[int]struct{})
+
+func LoadPrometheusExporterPortRangeMapping(configFp string) {
+	if Config == (HarvestConfig{}) {
+		LoadHarvestConfig(configFp)
+	}
+	exporters := *Config.Exporters
+	for k, v := range exporters {
+		if *v.Type == "Prometheus" {
+			if v.PortRange != nil {
```

Comment:
maybe pull the yaml unmarshaller from my branch so we don't need to parse here, makes this simpler

Reply:
done. with [2000-2030] parsing is failing. It works fine when port range is defined as (2000-2030)